sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosBulkOperationRequestAndResponse.java

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos;
+
+import com.azure.cosmos.util.Beta;
+import reactor.core.publisher.Flux;
+
+/**
+ * Request, response and the exception(if any) for a {@link CosmosItemOperation} request when processed using Bulk by calling
+ * {@link CosmosAsyncContainer#processBulkOperations(Flux, BulkProcessingOptions)}.
+ */
+@Beta(Beta.SinceVersion.V4_7_0)
+public final class CosmosBulkOperationRequestAndResponse<TContext> {
+
+    private final CosmosItemOperation operation;
+    private final CosmosBulkItemResponse response;
+    private final Throwable throwable;
+    private final TContext batchContext;
+
+    /**
+     * Initialises a new instance of {@link CosmosBulkOperationRequestAndResponse}.
+     *
+     * @param operation the {@link CosmosItemOperation} for which this response object has values.
+     * @param response the {@link CosmosBulkItemResponse} the bulk response.
+     * @param batchContext the context of this bulk request.
+     */
+    CosmosBulkOperationRequestAndResponse(CosmosItemOperation operation, CosmosBulkItemResponse response, TContext batchContext) {
+        this.operation = operation;
+        this.response = response;
+        this.throwable = null;
+        this.batchContext = batchContext;
+    }
+
+    /**
+     * Initialises a new instance of {@link CosmosBulkOperationRequestAndResponse}.
+     *
+     * @param operation the {@link CosmosItemOperation} for which this response object has values.
+     * @param throwable the {@link Throwable} for this request.
+     * @param batchContext the context of this bulk request.
+     */
+    CosmosBulkOperationRequestAndResponse(CosmosItemOperation operation, Throwable throwable, TContext batchContext) {
+        this.operation = operation;
+        this.response = null;
+        this.throwable = throwable;
+        this.batchContext = batchContext;
+    }
+
+    public CosmosItemOperation getOperation() {
+        return operation;
+    }
+
+    public CosmosBulkItemResponse getResponse() {
+        return response;
+    }
+
+    public Throwable getThrowable() {
```

Comment:
Do we have to keep this as Throwable - or can we make it CosmosException? When would it ever contain a Trhotwable whihc is not wrapped as CosmosException?

Reply:
Ideally we can keep it. But At most of the places I can see in try/catch we put Exception instead of CosmosException. So I am not sure if should do it?

Reply:
I think it would be better to translate to CosmosException or at least Exception.
there are very few errors which are Throwable but not Exception. (e.g. StackOverFlowError or OutofMemeryError). 

if you get a throwable which is not exception you should fail the whole operation rather than the individual request/response.

Reply:
Sure. Exception is cool. will do it. Thanks.